Docs: Added Experimental Edit pattern (#487)
carbon-design-system/carbon-website: src/pages/experimental/edit-pattern/index.mdx

---
title: Edit Pattern
---

<PageDescription>

Editing is making changes to a resource that already exists in the system.
Updating and editing should use the same behavior as when the resource is created. This should only vary in situations where the number of editable fields is significantly smaller or larger than during creation, or the context of current settings is required.

</PageDescription>

#### Status:

[Experimental](/experimental/overview/)

#### Maintainers:

[Vikki Paterson](https://github.com/vikkipaterson), [Chloe Poulter](https://github.com/chloepoulterdesign)

<AnchorLinks>

  <AnchorLink>Low-impact edit</AnchorLink>
  <AnchorLink>High-impact edit</AnchorLink>
  <AnchorLink>Optional extras</AnchorLink>

</AnchorLinks>

![Contextual image of an edit modal](images/01.png)

<Caption>Example of an edit modal in context</Caption>

## Low-impact edit

> Allow editing with no warning. Typically used when editing doesn’t have a wider impact.

Reviewer: ```suggestion
For low-impact edits, allow editing with no warning. This approach is typically used when editing doesn’t have a wider impact.
```

<Row>
<Column colLg={8}>

![Example of an edit button in a table](images/02.png)

<Caption>Example of an edit button in a table</Caption>

</Column>
</Row>

### Modal

If an object was created in a modal, its editable fields can be made available in a modal. A modal can also be used where a small subset of fields are editable.

> If enough fields are editable to require scrolling, use a side-panel or full-page edit dialog instead.

Reviewer: Did you want this to be two separate paragraphs? It looks like there is a line break, but no empty line between them so it'll render as its one paragraph.

Author: One paragraph is probably fine - I imagine it's the ol' trigger happy thumbs again

<Row>
<Column colLg={8}>

![Example of a low consequence edit action in a modal](images/03.png)

> <Caption>Example of a low conssequence edit action in a modal</Caption>

Reviewer: ```suggestion
<Caption>Example of a low consequence edit action in a modal</Caption>
```

</Column>
</Row>

### Side-panel

> If an object was created in a side panel, its editable fields can be made available in a side-panel. A side-panel can also be used where context in the main view of the screen is useful in the editing flow.

Reviewer: ```suggestion
If an object was created in a side panel, its editable fields can be made available in a side panel. A side panel can also be used where context in the main view of the screen is useful in the editing flow.
```

<Row>
<Column colLg={8}>

> ![Example of a low impact edit action in a side panel](images/04.png)

Reviewer: ```suggestion
![Example of a low-impact edit action in a side panel](images/04.png)
```

<Caption>Example of a low impact edit action in a side panel</Caption>

</Column>
</Row>

## High-impact edit

> Warn the user of the consequences of editing. Typically used when editing has a wider impact.

Reviewer: ```suggestion
Warn the user of the consequences of editing. This pattern is typically used when editing has a wider impact.
```

Inform users of the likely consequences of the edit to their system, and inform them that editing cannot be undone. Include a warning stage.

### Modal

> A high impact edit action of only a few fields can use a Danger modal. A second confirm edit modal should also be inserted before the changes are saved. If desired, the confirmation modal can include a [summary of changes](#summary-of-changes).

Reviewer: ```suggestion
A high-impact edit action of only a few fields can use a Danger modal. A second confirm edit modal should also be inserted before the changes are saved. If desired, the confirmation modal can include a [summary of changes](#summary-of-changes).
```

> A high-impact edit action of only a few fields can use a Danger modal. A second confirm edit modal should also be inserted before the changes are saved. If desired, the confirmation modal can include a [summary of changes](#summary-of-changes).

Reviewer: ```suggestion
A high-impact edit action of only a few fields can use a [danger modal]. A second confirm edit modal should also be inserted before the changes are saved. If desired, the confirmation modal can include a [summary of changes](#summary-of-changes).
```

<Row>
<Column colLg={8}>

![Example of a high-impact edit action in a side panel](images/06.png)

<Caption>Example of a high-impact edit action in a side panel</Caption>

</Column>
</Row>

High-impact side panel edits should also warn the user of destructive edits with a warning before changes are saved.

<Row>
<Column colLg={8}>

![Example of a warning screen in a side panel before a user saves changes](images/07.png)

<Caption>
  Example of a warning screen in a side panel before a user saves changes
</Caption>

</Column>
</Row>

### Full page

When an object was created in a full-page flow, or the volume of editable fields is high, use a full-page edit dialog.

<Row>
<Column colLg={8}>

![Example of a full-page edit view](images/08.png)

<Caption>Example of a full-page edit view</Caption>

</Column>
</Row>

<Row>
<Column colLg={8}>

![Example of a modal warning users of the consequences of saving after a full-page edit](images/09.png)

<Caption>
  Example of a modal warning users of the consequences of saving after a
  full-page edit
</Caption>

</Column>
</Row>

## Optional extras

### Summary of changes

> Adding a summary of changes to the warning stage of a High-impact flow can help the user manage and understand all of the changes they have made.

Reviewer: ```suggestion
Adding a summary of changes to the warning stage of a high-impact flow can help the user manage and understand all of the changes they have made.
```

<Row>
<Column colLg={8}>

![Example of a summary of changes in a modal](images/11.png)

<Caption>Example of a summary of changes in a modal</Caption>

</Column>
</Row>

### Non-editable fields

When a small number of fields are not editable, display their set value in a disabled field.

<Row>
<Column colLg={8}>

![Example of a single non-editable field](images/10.png)

<Caption>Example of a single non-editable field</Caption>

</Column>
</Row>

Where a majority of variables are not editable, consider using an option that will display only editable fields. In a small number of cases, make individual variables editable in modals.

### Success notifications

An optional success notification can be used to confirm that the user's changes have been saved.

<Row>
<Column colLg={8}>

> ![Example of a successful editing toaster notification](images/12.png)

Reviewer: ```suggestion
![Example of a successful editing toast notification](images/12.png)
```

> <Caption>Example of a successful editing toaster notification</Caption>

Reviewer: ```suggestion
<Caption>Example of a successful editing toast notification</Caption>
```
I assume this is supposed to say toast 😂

---
title: Edit Pattern
---

<PageDescription>

Editing is making changes to a resource that already exists in the system.

> Updating and editing should use the same behaviour as when the resource is created. This should only be varied in situations where the number of editable fields is significantly smaller or larger than during creation, or the context of current settings is required.

Reviewer: ```suggestion
Updating and editing should use the same behavior as when the resource is created. This should only vary in situations where the number of editable fields is significantly smaller or larger than during creation, or the context of current settings is required.
```
This isn't wrong, just switching to American spelling 😝

> Warn the user when the consequence of editing could damage resources or connections in their system.

Reviewer: ```suggestion
```

> Warn the user when the consequence of editing could damage resources or connections in their system.

Reviewer: According to our design team, we can't have line breaks in page descriptions. As we address consequences and related actions on each variant, we can probably delete this line. What do you think?

Author: Yes, I think we can cope with that

</PageDescription>

#### Status:

[Experimental](/experimental/overview/)

#### Maintainers:

[Vikki Paterson](https://github.com/vikkipaterson), [Chloe Poulter](https://github.com/chloepoulterdesign)

<AnchorLinks>

>   <AnchorLink>Low impact edit</AnchorLink>

Reviewer: ```suggestion
  <AnchorLink>Low-impact edit</AnchorLink>
```

>   <AnchorLink>High impact edit</AnchorLink>

Reviewer: ```suggestion
  <AnchorLink>High-impact edit</AnchorLink>
```

  <AnchorLink>Optional extras</AnchorLink>

</AnchorLinks>

![Contextual image of an edit modal](images/01.png)

<Caption>Example of an edit modal in context</Caption>

> ## Low impact edit

Reviewer: ```suggestion
## Low-impact edit
```

Allow editing with no warning. Typically used when editing doesn’t have a wider impact.

<Row>
<Column colLg={8}>

![Example of an edit button in a table](images/02.png)

<Caption>Example of an edit button in a table</Caption>

</Column>
</Row>

### Modal

If an object was created in a modal, its editable fields can be made available in a modal. A modal can also be used where a small subset of fields are editable.
If enough fields are editable to require scrolling, use a side-panel or full-page edit dialog instead.

<Row>
<Column colLg={8}>

![Example of a low consequence edit action in a modal](images/03.png)

<Caption>Example of a low conssequence edit action in a modal</Caption>

</Column>
</Row>

### Side-panel

If an object was created in a side panel, its editable fields can be made available in a side-panel. A side-panel can also be used where context in the main view of the screen is useful in the editing flow.

<Row>
<Column colLg={8}>

![Example of a low impact edit action in a side panel](images/04.png)

<Caption>Example of a low impact edit action in a side panel</Caption>

</Column>
</Row>

> ## High impact edit

Reviewer: ```suggestion
## High-impact edit
```

Warn the user of the consequences of editing. Typically used when editing has a wider impact.
Inform users of the likely consequences of the edit to their system, and inform them that editing cannot be undone. Include a warning stage.

### Modal

A high impact edit action of only a few fields can use a Danger modal. A second confirm edit modal should also be inserted before the changes are saved. If desired, the confirmation modal can include a [summary of changes](#summary-of-changes).

<Row>
<Column colLg={8}>

![Example of a high-impact editing modal and consequent warning modal](images/05.png)

<Caption>
  Example of a high-impact editing modal and consequent warning modal
</Caption>

</Column>
</Row>

### Side-panel

If an object was created in a side panel, or there are too many editable fields for a modal, a slide-over panel can be used.

<Row>
<Column colLg={8}>

> ![Example of a high impact edit action in a side-panel](images/06.png)

Reviewer: ```suggestion
![Example of a high-impact edit action in a side-panel](images/06.png)
```

> <Caption>Example of a high impact edit action in a side-panel</Caption>

Reviewer: ```suggestion
<Caption>Example of a high-impact edit action in a side-panel</Caption>
```

Author: I thought you'd inserted the hyphen into `side panel` in other reviews, but could be entirely mistaken

</Column>
</Row>

> High impact side panel edits should also warn the user of destructive edits with a warning before changes are saved.

Reviewer: ```suggestion
High-impact side panel edits should also warn the user of destructive edits with a warning before changes are saved.
```

<Row>
<Column colLg={8}>

![Example of a warning screen in a side panel before a user saves changes](images/07.png)

<Caption>
  Example of a warning screen in a side panel before a user saves changes
</Caption>

</Column>
</Row>

> ### Full-page

Reviewer: ```suggestion
### Full page
```

<Caption>Example of a full-page edit view</Caption>

</Column>
</Row>

<Row>
<Column colLg={8}>

![Example of a modal warning users of the consequences of saving after a full-page edit](images/09.png)

<Caption>
  Example of a modal warning users of the consequences of saving after a
  full-page edit
</Caption>

</Column>
</Row>

## Optional extras

### Summary of changes

Adding a summary of changes to the warning stage of a High-impact flow can help the user manage and understand all of the changes they have made.

<Row>
<Column colLg={8}>

![Example of a summary of changes in a modal](images/11.png)

<Caption>Example of a summary of changes in a modal</Caption>

</Column>
</Row>

### Non-editable fields

When a small number of fields are not editable, display their set value in a disabled field.

<Row>
<Column colLg={8}>

![Example of a single non-editable field](images/10.png)

<Caption>Example of a single non-editable field</Caption>

</Column>
</Row>

Where a majority of variables are not editable, consider using an option that will display only editable fields. In a small number of cases, make individual variables editable in modals.

### Success notifications

> An optional success notification can be used to confirm that the users' changes have been saved.

Reviewer: ```suggestion
An optional success notification can be used to confirm that the user's changes have been saved.
```

---

> title: Edit & Update Pattern

Reviewer: If I was looking for Update pattern guidance, I think I'd probably click on Edit. I think we can safely call it the "Edit pattern" and call out Update in the introduction you already do. It'll still appear upon search if a user can't find it.

---

<PageDescription>

> _Editing_ is making changes to a resource that already exists in the system.

Reviewer: ```suggestion
Editing is making changes to a resource that already exists in the system.
```

Updating and editing should use the same behaviour as when the resource is created. This should only be varied in situations where the number of editable fields is significantly smaller or larger than during creation, or the context of current settings is required.

Warn the user when the consequence of editing could damage resources or connections in their system.

</PageDescription>

#### Status:

[Experimental](/experimental/overview/)

#### Maintainers:

[Vikki Paterson](https://github.com/vikkipaterson), [Chloe Poulter](https://github.com/chloepoulterdesign)

> ### Options:

Reviewer: ```suggestion
### Options:
```
We can't have headers above the anchorlinks here. As you mention that the one should consider the consequences before allowing a deletion, I think we can just delete this header/line below.

Author: So just remove this line?

> Consider the consequences of editing before allowing it:

Reviewer: Same as comment above.
